src/realm/util/interprocess_mutex.hpp

```diff
@@ -260,11 +261,12 @@ inline void InterprocessMutex::set_shared_part(SharedPart& shared_part, const st
     m_lock_info = std::make_shared<LockInfo>();
     // Always use mod_Write to open file and retreive the uid in case other process
     // deletes the file.
-    m_lock_info->m_file.open(m_filename, File::mode_Write);
+    m_lock_info->m_file.open(m_filename, File::mode_Append);
     // exFAT does not allocate a unique id for the file until it's non-empty
     m_lock_info->m_file.resize(1);
     m_fileuid = m_lock_info->m_file.get_unique_id();
 
+    REALM_ASSERT_RELEASE(s_info_map->find(m_fileuid) == s_info_map->end());
```

Comment:
Are there some other places where this could be a problem? We do use open with truncate in a few other places.

Reply:
Haven't found anything anywhere else.

Reply:
This implies that the fix in https://github.com/realm/realm-core/pull/3883 was insufficient and that lines 267-268 are now unnecessary. Can you add some stronger checks on an InterprocessMutex to make sure that its file handle is allocated correctly on open?

Reply:
How it is so? Other fix was targeting different issue but essentially dealing with the same peculiarity.

'Append' mode doesn't resize newly created file (it just avoids truncating it on open and ultimately making uid available for other files). It's still needed to resize file to 1 to get real uid. 

Which stronger checks do you have in mind?

Reply:
I see, I misunderstood what the problem was here.
I was wondering if it might be possible to come up with a deterministic test case using InterprocessMutex and the SpawnedProcess test helper. But unless you can come up with something creative, I think it is just inherently racy. I suppose our existing multiprocess tests are actually hitting this and causing the hang which is reassuring, so technically it is covered.

Reply:
Ok, i've looked at it a bit again and added somewhat reliable test using just InterprocessMutex and spawn_process. Seems like we don't really have something like this at all. Should be beneficial. The logic is similar with how this is used in DB and it mimics almost the same pattern as with multiprocess tests from lang bindings test cases. It's still racy but somewhat reliably hits added assertions for me on exfat sdcard on macos, but needs a lot of iterations (>1000 sometimes) which takes tens of seconds. I've put just 10 to keep added time to milliseconds. If it hits an assertion or hangs on some run, we should be better equipped in the future to investigate as this time around. Would it be fine to merge?